internal/server/providers/azure_test.go

```diff
@@ -0,0 +1,328 @@
+			"visibility": null,
+			"onPremisesProvisioningErrors": []
+		},
+		{
+			"@odata.type": "#microsoft.graph.group",
+			"id": "ccc",
+			"deletedDateTime": null,
+			"classification": null,
+			"createdDateTime": "2022-06-07T20:40:20Z",
+			"creationOptions": [],
+			"description": null,
+			"displayName": "Developers",
+			"expirationDateTime": null,
+			"groupTypes": [],
+			"isAssignableToRole": null,
+			"mail": null,
+			"mailEnabled": false,
+			"mailNickname": "eee",
+			"membershipRule": null,
+			"membershipRuleProcessingState": null,
+			"onPremisesDomainName": null,
+			"onPremisesLastSyncDateTime": null,
+			"onPremisesNetBiosName": null,
+			"onPremisesSamAccountName": null,
+			"onPremisesSecurityIdentifier": null,
+			"onPremisesSyncEnabled": null,
+			"preferredDataLocation": null,
+			"preferredLanguage": null,
+			"proxyAddresses": [],
+			"renewedDateTime": "2022-06-07T20:40:20Z",
+			"resourceBehaviorOptions": [],
+			"resourceProvisioningOptions": [],
+			"securityEnabled": true,
+			"securityIdentifier": "qqq",
+			"theme": null,
+			"visibility": null,
+			"onPremisesProvisioningErrors": []
+		}
+	]
+}`
+
+func TestSyncAzureProviderUser(t *testing.T) {
+	server, ctx := setupOIDCTest(t)
+	serverURL := server.run(t)
+
+	db := setupDB(t)
+
+	provider := &models.Provider{
+		Name: "mock-azure",
+		Kind: models.AzureKind,
+	}
+
+	err := data.CreateProvider(db, provider)
+	assert.NilError(t, err)
+
+	oidc := &oidcImplementation{
+		ProviderID:   provider.ID,
+		Domain:       serverURL,
+		ClientID:     "whatever",
+		ClientSecret: "whatever",
+		RedirectURL:  "http://localhost:8301",
+	}
+	azure := &azure{OIDC: oidc}
+
+	now := time.Now().UTC()
+
+	claims := jwt.Claims{
+		Audience:  jwt.Audience([]string{server.knownClientID}),
+		NotBefore: jwt.NewNumericDate(now.Add(time.Minute * -5)), // adjust for clock drift
+		Expiry:    jwt.NewNumericDate(now.Add(time.Minute * 5)),
+		IssuedAt:  jwt.NewNumericDate(now),
+		Issuer:    serverURL,
+	}
+
+	body, err := testTokenResponse(claims, server.signingKey, "hello@example.com")
+	assert.NilError(t, err)
+
+	server.tokenResponse = tokenResponse{
+		code: 200,
+		body: body,
+	}
+
+	tests := []struct {
+		name         string
+		setupFunc    func(t *testing.T) *models.Identity
+		infoResponse string
+		verifyFunc   func(t *testing.T, err error, user *models.Identity)
+	}{
+		{
+			name: "invalid/expired access token is updated",
+			setupFunc: func(t *testing.T) *models.Identity {
+				user := &models.Identity{
+					Name: "sharrington@example.com",
+				}
+
+				err = data.CreateIdentity(db, user)
+				assert.NilError(t, err)
+
+				pu := &models.ProviderUser{
+					ProviderID: provider.ID,
+					IdentityID: user.ID,
+
+					Email:        user.Name,
+					RedirectURL:  "http://example.com",
+					AccessToken:  models.EncryptedAtRest("aaa"),
+					RefreshToken: models.EncryptedAtRest("bbb"),
+					ExpiresAt:    time.Now().Add(time.Minute * -5),
+					LastUpdate:   time.Now(),
```

Comment:
Maybe use a `LastUpdated` time in the last so that we can check that it was updated?